### PR TITLE
Use associated types

### LINKED ISSUE #240: Use associated types, remove old_orphan_check
Many type parameters — for example `T` on `Pixel<T>`, should actually be associated types. I suspect that it will fix the `old_orphan_check` warnings when implemented.

I am working on this, see https://github.com/ruud-v-a/image/tree/assoc-types.

### PR DESCRIPTION
**Pull request by ruud-v-a**
This still needs some cleanup, but this is going to be quite a big change, so I’m posting it here for early feedback. This resolves #240. There are a lot of breaking changes, so we might want to bump the version number when this can be merged.

- This uses associated types where appropriate. The most important changes are that the type parameters of `Pixel` and `GenericImage` are now associated types. `ImageBuffer` requires less type params.
- This resolves the `old_orphan_check` and `old_impl_check` warnings.

Associated types feel like the “right” way to go. In many places it simplifies the API, but naming the subpixel type has become uglier: for `I: GenericImage`, the subpixel type is `<I::Pixel as Pixel>::Subpixel`, instead of a `T` param that was often used before. However, the `T` param was often unconstrained, and there is no need for it with associated types.

A few issues:
- The return type of `into_dynamic` cannot be `ImageBuffer` any more. I removed the function.
- Many functions now require a `'static` bound on `I: GenericImage`. How will this impact users of the library, and is there a better way to do it?
- Should `Pixel::Subpixel` and `GenericImage::Pixel` have `'static` bounds? I had hoped that adding those would enable omitting the `where P::Subpixel: 'static` in most places, but unfortunately this is not the case.
- The pixel type was sometimes allowed to be unsized. What is the use of that?

### CODE REVIEW

**Comment by nwin** [comment]
The Sized bound can go. It is just an artifact introduced during development.

File: src/buffer.rs [hunk 1]
```diff
@@ -362,29 +365,14 @@ where Container: ArrayLike<T>, T: Primitive + 'static, PixelType: Pixel<T> + 'st
...
-    #[experimental]
-    pub fn into_dynamic(self) -> ImageBuffer<Container, T, &'static Any> {
-        ImageBuffer {
-            data: self.data,
-            width: self.width,
-            height: self.height,
-            type_marker: self.type_marker
-
-        }
-    }
```
Comment (nwin) [inline src/buffer.rs]: This was useless anyway since I did not implement the accompanying code yet. But I think it should work when implementing Pixel for Any…

File: src/buffer.rs [hunk 2]
```diff
@@ -406,54 +395,62 @@ where Container: ArrayLike<T>, T: Primitive + 'static, PixelType: Pixel<T> + 'st
...
 // concrete implementation for `Vec`-baked buffers
-impl<T, PixelType> ImageBuffer<Vec<T>, T, PixelType>
-where T: Primitive + 'static, PixelType: Pixel<T> + 'static {
-    /// Creates a new image buffer based on a `Vec<T>`.
-    pub fn new(width: u32, height: u32) -> ImageBuffer<Vec<T>, T, PixelType> {
+// TODO: I think that rustc does not "see" this impl any more: the impl with
+// Container meets the same requirements. At least, I got compile errors that
+// there is no such function as `into_vec`, whereas `into_raw` did work, and
+// `into_vec` is redundant anyway, because `into_raw` will give you the vector,
+// and it is more generic.
```
Comment (nwin) [inline src/buffer.rs]: I'm a bit worried about that. Is this a compiler bug or something I did wrong? Does `rustc` still find `new`? If not this is bad. How is this supposed to work? Afaik one can’t just add a generic constructor to the ArrayLike trait because static methods are not object safe. 

PS: I just added `into_vec`  etc. to match the stdlib.

File: src/buffer.rs (hunk 1 above)
Comment (ruud-v-a) [inline src/buffer.rs]: Yes, in that case it would work.

File: src/buffer.rs (hunk 2 above)
Comment (ruud-v-a) [inline src/buffer.rs]: I don’t know yet whether it does not break anything, building in debug mode causes an ICE, but at least the library compiles in release mode, so `new` still works. I think that we should not have two identical methods, but removing `into_vec` does not belong in this PR either way.

Comment (nwin) [inline src/buffer.rs]: Did you file a bug to rustc about that?

Comment (ruud-v-a) [inline src/buffer.rs]: The report is [here](https://github.com/rust-lang/rust/issues/20797). I’m not working around it, it’s just that it prevents me from running the test suite until it is fixed.

Comment (nwin) [inline src/buffer.rs]: I see… :/

**Comment by TyOverby** [comment]
I was just about to start writing a change that was very similar to this.  Is there an update on being able to build the library?  I'd very much like to see this merged!

**Comment by ruud-v-a** [comment]
I’m sorry, I have been away for a few days. The blocking issue on rustc has been closed, but there were some breaking changes as well. I’ll try to get it to compile again as soon as possible.

**Comment by ruud-v-a** [comment]
This compiles with the latest nightly, and the test suite compiles and passes as well. There is still a bit of cleanup to be done.

**Comment by nwin** [comment]
Great work! I'll merge it for now, the cleanup we could do later.

**Comment by bvssvni** [comment]
@ruud-v-a You're awesome!
